packages/vite/src/node/plugins/html.ts

```diff
@@ -850,24 +861,35 @@ export function postImportMapHook(): IndexHtmlTransformHook {
 
 export function resolveHtmlTransforms(
   plugins: readonly Plugin[]
-): [IndexHtmlTransformHook[], IndexHtmlTransformHook[]] {
+): [
+  IndexHtmlTransformHook[],
+  IndexHtmlTransformHook[],
+  IndexHtmlTransformHook[]
+] {
   const preHooks: IndexHtmlTransformHook[] = []
+  const normalHooks: IndexHtmlTransformHook[] = []
   const postHooks: IndexHtmlTransformHook[] = []
 
   for (const plugin of plugins) {
     const hook = plugin.transformIndexHtml
-    if (hook) {
-      if (typeof hook === 'function') {
-        postHooks.push(hook)
-      } else if (hook.enforce === 'pre') {
-        preHooks.push(hook.transform)
+    if (!hook) continue
+
+    if (typeof hook === 'function') {
+      normalHooks.push(hook)
+    } else {
+      const order = hook.order ?? hook.enforce
+      const handler = hook.handler ?? hook.transform
+      if (order === 'pre') {
+        preHooks.push(handler)
+      } else if (order === 'post') {
+        postHooks.push(handler)
       } else {
-        postHooks.push(hook.transform)
+        normalHooks.push(handler)
       }
```

Comment:
I think this is a breaking change. Though, I'm not sure how we can avoid this.

For example:
```js
const plugins = [
  { name: 'a', transformIndexHtml: { enforce: 'post', transform: () => { console.log('a') } } },
  { name: 'b', transformIndexHtml: () => { console.log('b') } }
]

// before
const hooks = [
  () => { console.log('a') }, // postHooks
  () => { console.log('b') } // postHooks
]

// after
const hooks = [
  () => { console.log('b') }, // normalHooks
  () => { console.log('a') } // postHooks
]
```

Reply:
Good catch! I like the idea of `enforce` (with two values, and `post` by default) to `order` (with three values, the same as with the regular plugins).
Maybe we should convert `enforce: 'post'` and `enforce: undefined` to `order: post` and we make the new default for `order: undefined` work like in this PR (still a post hook, but before all 'post' hooks)

Reply:
I feel that we should align its behavior with the rest of the hooks. Where we could consider Vite's transformation as a core plugin. Having this would allow plugins to set `post` to move after normal hooks. I don't really consider this a breaking change as it follows the intuitive and aligns with other hooks and plugin order.

Reply:
Yeah, I think it's better to ailgn with the rest of the hooks.

I've come up with an idea now.
How about mapping `enforce: 'post'` to `normalHooks`? (`order: 'post'` is stilll mapped to `postHooks`)
The migration will be counterintuitive, but we can avoid this behavior change.

|                      | before| after|
|----------------------|-------|------|
| no `enforce`/`order` |    2  |    2 |
| `enforce: 'pre'`     |    1  |    1 |
| `enforce: 'post'`    |    2  |    2 |
| `order: 'pre'`       |    -  |    1 | 
| `order: 'post'`      |    -  |    3 |

```js
const plugins1 = [
  { name: 'a', transformIndexHtml: { enforce: 'post', transform: () => { console.log('a') } } },
  { name: 'b', transformIndexHtml: () => { console.log('b') } }
]
const plugins2 = [
  { name: 'a', transformIndexHtml: { order: 'post', transform: () => { console.log('a') } } },
  { name: 'b', transformIndexHtml: () => { console.log('b') } }
]

// before
const hooks1 = [
  () => { console.log('a') }, // postHooks
  () => { console.log('b') } // postHooks
]

// after
const hooks1 = [
  () => { console.log('a') }, // normalHooks
  () => { console.log('b') } // normalHooks
]
const hooks2 = [
  () => { console.log('b') }, // normalHooks
  () => { console.log('a') } // postHooks
]
```

Reply:
Would you explain a bit more about your proposal @sapphi-red?

> I don't really consider this a breaking change as it follows the intuitive and aligns with other hooks and plugin order.

Before we only had two values, and the default was `'post'`, so it is somewhat of a breaking change. That said, I doubt it will break any code.

> I feel that we should align its behavior with the rest of the hooks. Where we could consider Vite's transformation as a core plugin. Having this would allow plugins to set post to move after normal hooks.

@antfu I also would like the same. Since we are doing this change, maybe we could review the default for normal plugins? Thinking about the transforms that Vite does as a plugin, I think that `order: undefined` should go before them, and not after. I always found it strange to have to add `enforce: pre` to add a script for example... I think normal plugins should go before the main Vite pipeline (especially during build)

What about:
- `enforce: 'pre'` -> `order: 'pre'`
- `enforce: undefined` -> `order: 'post'`
- `enforce: 'post` -> `order: 'post'` 

Hooks order:
`order: 'pre'` => `order: undefined` (normal) => Vite internal => `order: post`

This isn't a breaking change, since `enforce` will wire `order` in the exact same way we have now.

Reply:
I think @sapphi-red are proposing the same idea to you, by treating `enforce` and `order` differently. I agree this would indeed make the behavior strictly the same as previous, but my question is if that really affects any real-world usage or if we are over thinking?

I am not sure if any plugin would use `enforce: undefined` as in that case they can directly use function hook instead of object hook. And even with `enfore: post`, we don't guarantee the exact order of each plugin, but that relies on how user-defined it. I would think a good plugin should not have too much constraint on the order of two `post` hooks.

Reply:
I think my idea is different from @patak-dev's one.

- `enforce: 'pre'` -> `order: 'pre'`
- no `enforce`(directly using a function) / `enforce: undefined` -> `order: undefined` (normal)
- `enforce: 'post'` -> `order: undefined` (normal)

To make it clear, my proposal's implementation will be:
```ts
export function resolveHtmlTransforms(
  plugins: readonly Plugin[]
): [
  IndexHtmlTransformHook[],
  IndexHtmlTransformHook[],
  IndexHtmlTransformHook[]
] {
  const preHooks: IndexHtmlTransformHook[] = []
  const normalHooks: IndexHtmlTransformHook[] = []
  const postHooks: IndexHtmlTransformHook[] = []

  for (const plugin of plugins) {
    const hook = plugin.transformIndexHtml
    if (!hook) continue

    if (typeof hook === 'function') {
      normalHooks.push(hook)
    } else {
      const order = hook.order ?? hook.enforce
      // @ts-expect-error union type
      const handler = hook.handler ?? hook.transform

      if (order === 'pre') {
        preHooks.push(handler)
      } else if (order === 'post') {
        // CHANGE START
        if (hook.order === undefined) {
          normalHooks.push(handler) // treat `enforce: 'post'` as normal hook
        } else {
          postHooks.push(handler)
        }
        // CHANGE END
      } else {
        normalHooks.push(handler)
      }
    }
  }

  return [preHooks, normalHooks, postHooks]
}
```
Also I'm not changing the hooks order:
`order: 'pre'` => Vite internal => `order: undefined`(normal) => `order: post`

So this is renaming the current post hook as normal hook and introducing a new post hook.

IIUC @patak-dev's idea is to introduce a new normal hook.
I think your one is still a breaking change because when using function directly, it will change from post hook to normal hook.

> my question is if that really affects any real-world usage or if we are over thinking?

If two plugins (one is using function directly and one is `enforce: 'post'`) are injecting script tags, this will be a problem because the order of injected script matters. Though, I'm not sure if there's a real-world usage of this.

Reply:
Now I see that my proposal to change the default for the new `order` to be before the internal Vite hook isn't possible, sorry for the noise. And I think I understand now why you were fine with the small possibility of a breaking change.

But looks like @sapphi-red's proposal would work. Both `enforce: undefined` (or function form) and `enforce: post` are mapped to the same value (that is what I was trying to propose), so there is no breaking change. I think we should go with it 👍🏼

Reply:
While I agree that @sapphi-red's proposal could be a solution, I am just not sure if the complexity is necessary here. As the difference is so subtle (and technically any change could be considered breaking if others rely on the behavior), maybe we could run a ecosystem CI and see if it affects any known integration before making the change?